### DISCUSSION TITLE
Debug suggestion for PICO Platformio+VS on Linux

### DISCUSSION THREAD

**Discussion by git2212** [earlephilhower/arduino-pico#1424] [General]
Hello,

I've been successful using the RP2040 based tiny pico in Platformio using the config:

platform = https://github.com/maxgerhardt/platform-raspberrypi.git
board_build.core = earlephilhower
framework = arduino

So far so good ... now looking at being able to reliably debug sketches from within the Platformio+VS environment. From what I understand so far the PICO could use the second core for USB CMSIS-DAP based debugging. Like the idea of fewer anything that can go wrong so this seems very appealing. My currently supported upload options (as per the message in platformio) do not include pico-debug. While picoprobe is supported, it seems to require external hardware, etc.

As I am a rusty amateur on a good day, I am none the wiser after reading [https://github.com/majbthrd/pico-debug/blob/master/howto/README.md](https://github.com/majbthrd/pico-debug/blob/master/howto/README.md) Is there a way to set up pico-debug for my given dev environment on Linux? The ideal setup would run reliably and work with the USB cable (already using for uploads).

Thanks in advance :)

**Comment by maxgerhardt** [comment]
If the README is right and just have to upload the .uf2 to on it and it'll show up as CMSIS-DAP debugger (and this is somehow additive to the USB devices already created by the firmware), you'd just need to use 
```ini
debug_tool = cmsis-dap
```
to configure PlatformIO to use it. I have not tried this yet.

**Edit:** Ah, so definitely don't use USB in your firmware.

>the USB peripheral is used to provide the debugger, so the user code cannot use it as well

So per https://arduino-pico.readthedocs.io/en/latest/platformio.html you'll need the NO_USB configuration.

**Comment by git2212** [comment]
So here's the current state:

Downloaded the u2f files from [https://github.com/majbthrd/pico-debug/releases](https://github.com/majbthrd/pico-debug/releases).
Put the board in bootloader mode then using 

`~/.platformio/packages/tool-rp2040tools/picotool load -v -x "~/Downloads/pico-debug-maxram.u2f"`

 I uploaded pico-debug to the board... I can confirm that this worked as dmesg returned:

```
[649792.976255] usb 1-1.1.3.2: new full-speed USB device number 10 using ehci-pci
[649793.189865] usb 1-1.1.3.2: New USB device found, idVendor=1209, idProduct=2488, bcdDevice=10.05
[649793.189869] usb 1-1.1.3.2: New USB device strings: Mfr=2, Product=1, SerialNumber=3
[649793.189871] usb 1-1.1.3.2: Product: CMSIS-DAP
[649793.189873] usb 1-1.1.3.2: Manufacturer: pico-debug
[649793.189875] usb 1-1.1.3.2: SerialNumber: 4250304232373612
[649793.191370] hid-generic 0003:1209:2488.0020: hiddev1,hidraw3: USB HID v1.11 Device [pico-debug CMSIS-DAP] on usb-0000:00:1a.0-1.1.3.2/input0

```
... so the board no longer came up as a USB device :)

Then created an empty platformio project, and set the following options in the ini file:

```
[env:pico-debug]
board=pico
framework = arduino
board_build.f_cpu = 133000000L

platform = https://github.com/maxgerhardt/platform-raspberrypi.git
board_build.core = earlephilhower

upload_protocol = cmsis-dap
debug_tool = cmsis-dap
debug_init_break = tbreak setup

build_type = debug
build_flags = 
    -O0
    -D USE_TINYUSB
    -D PIO_FRAMEWORK_ARDUINO_NO_USB
```
Since I am running on linux I had to run the following command as well:

`sudo apt install libhidapi-dev`

Then compile ... all good ... upload however comes back with an error...
```
CURRENT: upload_protocol = cmsis-dap
Uploading .pio/build/generic/firmware.elf
Open On-Chip Debugger 0.11.0-g8e3c38f78-dirty (2023-02-16-01:21)
Licensed under GNU GPL v2
For bug reports, read
        http://openocd.org/doc/doxygen/bugs.html
debug_level: 1

adapter speed: 1000 kHz

in procedure 'program'
** OpenOCD init failed **
shutdown command invoked

*** [upload] Error 1
```
Just out of curiousity pressed F5, debugger seemed to start but not much appears to be happening...

So not sure if I missed a step or there is something particular to my platform/setup?

**Reply by maxgerhardt** [reply]
It's possible that OpenOCD just can't access the device because you're missing the [udev rules](https://docs.platformio.org/en/latest/core/installation/udev-rules.html#platformio-udev-rules). If you have them already installed (and added your user to the right groups and rebooted per documentation), it may be possible that the rules don't yet cover the PID+VID of the Pico-Debug device specifically, so you could maybe expand them with a `idVendor=1209, idProduct=2488` match.

**Reply by git2212** [reply]
So I had a generic entry in the udev rules installed by platformio as follows:

`ATTRS{product}=="*CMSIS-DAP*", MODE="0666", ENV{ID_MM_DEVICE_IGNORE}="1", ENV{ID_MM_PORT_IGNORE}="1"`

What should I put in there again? ... perhaps this?

`ACTION!="add|change", ATTRS{idVendor}=="1209" , ATTRS{product}=="2488", MODE="666" GROUP="plugdev", ENV{ID_MM_DEVICE_IGNORE}="1", ENV{ID_MM_PORT_IGNORE}="1"`

Also, how could I verify that the OpenOCD installed in 

is compatible?... according to this [https://github.com/majbthrd/pico-debug/blob/master/howto/openocd.md](https://github.com/majbthrd/pico-debug/blob/master/howto/openocd.md) there could be a compatibility issue?

a more detailed output from the upload command seems to indicate that it's looking to activate 'swd' 

```
CURRENT: upload_protocol = cmsis-dap
openocd -d2 -s /home/arthur/.platformio/packages/tool-openocd-rp2040-earlephilhower/share/openocd/scripts -f interface/cmsis-dap.cfg -f target/rp2040.cfg -c "adapter speed 1000" -c "program {.pio/build/generic/firmware.elf}  verify reset; shutdown;"
Open On-Chip Debugger 0.11.0-g8e3c38f78-dirty (2023-02-16-01:21)
Licensed under GNU GPL v2
For bug reports, read
        http://openocd.org/doc/doxygen/bugs.html
debug_level: 2

Info : auto-selecting first available session transport "swd". To override use 'transport select <transport>'.
Info : Hardware thread awareness created
Info : Hardware thread awareness created
Info : RP2040 Flash Bank Command
adapter speed: 1000 kHz

Info : CMSIS-DAP: SWD  Supported
Info : CMSIS-DAP: FW Version = 2.0.0
Info : CMSIS-DAP: Serial# = 4250304232373612
Info : CMSIS-DAP: Interface Initialised (SWD)
Info : SWCLK/TCK = 0 SWDIO/TMS = 0 TDI = 0 TDO = 0 nTRST = 0 nRESET = 0
Info : CMSIS-DAP: Interface ready
Info : clock speed 1000 kHz
Info : SWD DPIDR 0x0bc12477
Info : SWD DLPIDR 0x00000001
Info : DAP init failed
in procedure 'program'
** OpenOCD init failed **
shutdown command invoked

*** [upload] Error 1

```
is this right since we are not actually connecting to the swd pins?

**Reply by earlephilhower** [reply]
Pico-Debug is not a normal CMSIS-DAP board.  You need to do `source [find board/pico-debug.cfg]` which will source a CMSIS-DAP with the special, 1-core config required.

If I were you, I'd spend the $4 on another Pico, flash it with the free Picoprobe FW, add a SWD header to your target board, and just use that.  Pico-Debug is a heck of a neat hack, but does require jumping through hoops and limits your sketch (RAM limited, no USB, no 2nd core).  Picoprobe is basically painless and very widely used.

**Reply by git2212** [reply]
sorry, I meant OpenOCD in  
`.platformio/packages/tool-openocd-rp2040-earlephilhower`

**Reply by git2212** [reply]
> Pico-Debug is not a normal CMSIS-DAP board. You need to do `source [find board/pico-debug.cfg]` which will source a CMSIS-DAP with the special, 1-core config required.
> 
> If I were you, I'd spend the $4 on another Pico, flash it with the free Picoprobe FW, add a SWD header to your target board, and just use that. Pico-Debug is a heck of a neat hack, but does require jumping through hoops and limits your sketch (RAM limited, no USB, no 2nd core). Picoprobe is basically painless and very widely used.

I hear you and agree... I do have other pico's and not trying to be cheap lol... have to use these small boards for specific projects and they don't break out the SWD pins as far as I know... and in that case, I have to jump through the hoops lol. what does this `source [find board/pico-debug.cfg]` have to look like and where to put it?

**Reply by git2212** [reply]
now I am so close that it's almost a matter of professional pride to get it working lol... and despite having flashbacks to my assembler debug days in the 80ies, I would consider myself an amateur in this round :) ... hence all my silly questions...

**Reply by earlephilhower** [reply]
`openocd -c 'source [find board/pico-debug.cfg]'` would be one way to call it.  (-c == execute command IIRC)

**Reply by git2212** [reply]
looking at the upload command that platfomio generates, that's quite a difference to be sure

`openocd -d2 -s /home/arthur/.platformio/packages/tool-openocd-rp2040-earlephilhower/share/openocd/scripts -f interface/cmsis-dap.cfg -f target/rp2040.cfg -c "adapter speed 1000" -c "program {.pio/build/generic/firmware.elf}  verify reset; shutdown;"
`
as I understand it the pico-debug.cfg is part of the OpenOCD which is compatible with the pico-debug image... I can't find that file anywhere in `.platformio/packages/tool-openocd-rp2040-earlephilhower` so I am guessing the openOCD version there is not compatible... correct?

**Reply by earlephilhower** [reply]
If you're using my version of the tool chain, it's definitely in there in the share directory.

In any case, the command I listed only sets the board properly, it doesn't do an upload or debug.  You'll need to figure that out yourself, combining the two.

**Reply by maxgerhardt** [reply]
Hmm actually I'm a bit skeptical about the "pico-debug is not a CMSIS-DAP" because the `boards/pico-debug.cfg` is literally
```
source [find interface/cmsis-dap.cfg]
adapter speed 4000

set CHIPNAME rp2040
source [find target/rp2040-core0.cfg]
```
but testing it out now.

**Reply by maxgerhardt** [reply]
Yeah but I guess it also doesn't harm it when using the dedicated config file for it, that at least has the default speed.

![grafik](https://user-images.githubusercontent.com/26485477/236462912-d124401a-f207-44d6-aaef-e288e99224fe.png)

Works fine after adding the necessary code and builder script additions into PlatformIO, will file PR soon.

Just wondering though: With the two pico-debug variants of gimmecache (will load itself into upper SRAM) and MAXRAM (will load itself into the flash cache), I've had no success yet using the MAXRAM variant. Is the SDK or this core using that?

**Reply by maxgerhardt** [reply]
PR pending in https://github.com/earlephilhower/arduino-pico/pull/1427.

**Comment by maxgerhardt** [comment]
Ah then it'll just need array entry in the forked platform-raspberrypi and an addition in the boards array -- will look into tomorrow

**Reply by earlephilhower** [reply]
FWIW there are a couple build options (`-DNO_USB -DDISABLE_USB_SERIAL` and build.ram_length=240k) needed during build as well.

**Comment by git2212** [comment]
Hello Gents, just in case it gets lost in the shuffle, I've tried the solution proposed and was not successful, see post for [#1427](https://github.com/earlephilhower/arduino-pico/pull/1427)

**Reply by git2212** [reply]
Ok, good news. After I noticed that my .platformio/platforms/raspberrypi/boards/adafruit_feather_can.json was missing, my platformio VS plugin was clearly not pulling down the right files... I deleted everything in .platformio/platforms/raspberrypi which forced a fresh download of everything from git after which compile, upload and debug works.
You guys are quick and responsive, in the old days I would have sent you a bottle of good scotch... these days we'll have to settle for bubble tea or something :)
